# Практическая работа №1: Исследование алгоритмов формирования аддитивных цепочек

Выполнил студент гр. 0303 Калмак Даниил, вариант нечетный.

## Цель работы

Формирование представления о аддитивных цепочках, выработать уме-
ние составлять и применять алгоритмы для нахождения минимальных
аддитивных цепочек для заданного числа, привить навык использования
систем компьютерной математики для реализации алгоритмов.

## Основные теоретические положения

Приводятся основные теоретические положения для данной практической работы: определения, формулы в стиле $ \LaTeX $ etc. **Данный пункт является обязательным!**

При оформлении формул $ \LaTeX $ просьба не использовать русские буквы, поскольку возможно некорректное отображение при загрузке на Github.

## Постановка задачи

Реализовать точные и приближённые алгоритмы нахождения минимальных аддитивных цепочек с использованием системы компьютерной математики SageMath, провести анализ алгоритмов. Полученные результаты
содержательно проинтерпретировать.

## Выполнение работы


1. Вручную (т.е. не реализовывая алгоритм на Sage) была построена последовательность вычислений бинарным методом и методом множителей для $x^{30}, x^{45}$. Были сравнены количества операций для каждого метода и сделаны выводы.

n = 30. Бинарный метод:
Бинарная запись числа: 11110. Отбрасываем старший бит и остается 1110. Заменяем 1 на SX, а 0 на X, где S - операция возведения в квадрат, а X - умножение на x. SXSXSXS
$x, x^2, x^3, x^6, x^7, x^{14}, x^{15}, x^{30}$. 

Итого 7 операций.

Метод множителей:
30 = 2 * 15

$x^2 = x * x$ - одна операция

$y^{15} = (y^3)^5 = (y^3)^4 * y^3 = ((y^3)^2)^2 * y^3, y^3 = y^2*y$ - 5 операций

Всего 6 операций.

n = 45. Бинарный метод:
Бинарная запись числа: 101101. Отбрасываем старший бит и остается 01101. Заменяем 1 на SX, а 0 на X, где S - операция возведения в квадрат, а X - умножение на x. SSXSXSSX
$x, x^2, x^4, x^5, x^{10}, x^{11}, x^{22}, x^{44}, x^{45}$. 

Итого 8 операций.

Метод множителей:
45 = 3 * 15

$x^3 = x * x * x$ - две операции

$y^{15} = (y^3)^5 = (y^3)^4 * y^3 = ((y^3)^2)^2 * y^3, y^3 = y^2*y$ - 5 операций

Всего 7 операций.

Таким образом, метод множителей оказался выгоднее в обоих случаях. Однако такое происходит не всегда, например, n = 33.

2. Был реализован алгоритм Брауэра для вычисления приближённых аддитивных цепочек для различных чисел при варьировании 1 параметра 𝑘, сопоставлены длины полученных аддитивных цепочек с минимальной аддитивной цепочкой для заданного числа. Сделаны выводы.

In [156]:
import time

def Brauer(n, k):
    d = 2^k
    q = [] # список для целых частей от деления на d
    r = [] # список для остатков от деления на d
    n_brauer = n
    start_time = time.time()
    while n_brauer>=d:
        q.append(n_brauer // d) # добавление целых частей
        r.append(n_brauer % d) # добавление остатков
        n_brauer = q[-1]
    #print(q)
    #print(r)
    answer = []
    for i in q:
        if i < d: # если целая часть меньше d, то добавляем в цепочку числа от 1 до d-1
            answer.extend(range(1, d))
        i_1 = i
        while i_1 <= d*i: # пока целая часть меньше или равна d, удваиваем целую часть и добавляем
            answer.append(i_1)
            i_1 = i_1 * 2
    answer = sorted(set(answer))
    answer.append(answer[-1]+r[0])
    print(f"B_k({n}) = {answer}")
    print("Длина цепочки: ", len(answer))
    print("Время выполнения: {} seconds".format(time.time()-start_time))
    
Brauer(30, 3)
Brauer(45, 1)
Brauer(45, 3)
Brauer(45, 5)
test = [1023, 1024, 1057, 1538, 2000]
for i in test:
    print("-----")
    Brauer(i, 3)

B_k(30) = [1, 2, 3, 4, 5, 6, 7, 12, 24, 30]
Длина цепочки:  10
Время выполнения: 0.0003542900085449219 seconds
B_k(45) = [1, 2, 4, 5, 10, 11, 22, 44, 45]
Длина цепочки:  9
Время выполнения: 0.0001087188720703125 seconds
B_k(45) = [1, 2, 3, 4, 5, 6, 7, 10, 20, 40, 45]
Длина цепочки:  11
Время выполнения: 8.797645568847656e-05 seconds
B_k(45) = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 45]
Длина цепочки:  33
Время выполнения: 7.05718994140625e-05 seconds
-----
B_k(1023) = [1, 2, 3, 4, 5, 6, 7, 8, 15, 30, 60, 120, 127, 254, 508, 1016, 1023]
Длина цепочки:  17
Время выполнения: 7.486343383789062e-05 seconds
-----
B_k(1024) = [1, 2, 3, 4, 5, 6, 7, 8, 16, 32, 64, 128, 256, 512, 1024, 1024]
Длина цепочки:  16
Время выполнения: 7.343292236328125e-05 seconds
-----
B_k(1057) = [1, 2, 3, 4, 5, 6, 7, 8, 16, 32, 64, 128, 132, 264, 528, 1056, 1057]
Длина цепочки:  17
Время выполнения: 7.200241088867188e-05 seconds
-----
B_

Для n = 30 длина цепочки 10 при k = 3, что больше минимальной цепочки равной 5. Для n = 45 длина цепочки 11 при k = 3, что больше минимальной цепочки равной 6. Для разных k получаются разные длины цепочек, причем время приблизительно одинаковое.

3. Реализован алгоритм дробления вектора индексов для нахождения минимальной звёздной цепочки для заданного числа. Протестирован алгоритм для 5 значений 𝑛 > 1000. Указать, сколько времени потребовалось на поиск цепочки и какая цепочка получилась. Сравнить с предыдущими методами, сделать выводы.

In [157]:
def nu(n): #вес Хемминга
    n = Integer(n)
    return n.popcount()

In [158]:
def _lambda(n):
    res = 0
    while n != 0:
        res += 1
        n //= 2
    return res - 1

In [159]:
def star_chain(vec): # преобразование цепочки из вектора
    chain = [1]
    for i in range(len(vec)):
        chain.append(chain[i] + chain[vec[i] - 1])
    return chain

In [160]:
def check(vec): # проверка для следующего вектора
    for i in vec:
        if i > 1:
            return False
    return True

In [161]:
def next_vector(vec): # следующий вектор
    l = len(vec)
    for i in range(l):
        if vec[l - i - 1] > 1:
            vec[l - i - 1] -= 1
            for j in range(l - i, l):
                vec[j] = j + 1
            return vec

In [176]:
def splitting_vectors(n): # алгоритм дробления вектора индексов
    l_lower = math.ceil(math.log2(n)) # нижняя граница
    l_upper = int(math.log2(n) + nu(n) - 1) # верхняя граница
    for m in range(l_lower, l_upper + 1):
        q = m // 2
        fix = [i for i in range(1, q + 1)]
        
        while true: # перебор фиксированных частей
            change = [i for i in range(q + 1, m + 1)]
            a = star_chain(fix)[-1]
            a_min = a + m - q
            a_max = a * 2 ** (m - q)

            if n < a_min or n > a_max:
                if check(fix) and len(fix) > 1:
                    break
                fix = next_vector(fix)
                continue
                
            while true: # перебор меняющихся частей
                if star_chain(fix + change)[-1] == n:
                    return fix + change
                if check(change):
                    break
                change = next_vector(change)
            if check(fix):
                break
            fix = next_vector(fix)   
    print("Не найдено!")

In [177]:
#n = int(input())
n = [30, 45]
for i in n:
    start_time = time.time()
    res = splitting_vectors(i)
    print("Время выполнения: {} seconds".format(time.time()-start_time))
    print("Вектор: ", res)
    print("Цепочка: ", star_chain(res))

Время выполнения: 0.0008275508880615234 seconds
Вектор:  [1, 2, 2, 4, 5, 4]
Цепочка:  [1, 2, 4, 6, 12, 24, 30]
Время выполнения: 0.0048046112060546875 seconds
Вектор:  [1, 2, 1, 4, 5, 6, 4]
Цепочка:  [1, 2, 4, 5, 10, 20, 40, 45]


In [178]:
# test = [1023, 1024, 1057, 1538, 2000]
# for i in test:
#     print("-----")
#     print(i)
#     start_time = time.time()
#     res = splitting_vectors(i)
#     print("Время выполнения: {} seconds".format(time.time()-start_time))
#     print("Вектор: ", res)
#     print("Цепочка: ", star_chain(res))

|      | vector | chain | time  |
|------|--------|-------|-------|
| 1023 |    1, 1, 3, 4, 5, 4, 7, 8, 9, 10, 11, 8, 3    |    1, 2, 3, 6, 12, 24, 30, 60, 120, 240, 480, 960, 1020, 1023   | 268.4 |
| 1024 |    1, 2, 3, 4, 5, 6, 7, 8, 9, 10    |    1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024   | 0.00005  |
| 1057 |    1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 6, 1    |   1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1056, 1057    | 1.47  |
| 1538 |    1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 10, 2    |   1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1536, 1538    | 0.09  |
| 2000 |    1, 2, 3, 4, 4, 1, 7, 8, 9, 10, 11, 12, 11    |   1, 2, 4, 8, 16, 24, 25, 50, 100, 200, 400, 800, 1600, 2000    | 11.35 |

Алгоритм дробления вектора индексов работает дольше, однако дает цепочки достаточно минимальные.

4. Была проверена гипотеза Шольца–Брауэра для всех натуральных 1 ⩽ 𝑛 ⩽ 12 на алгоритме дробления вектора индексов. Результаты оформлены в виде таблицы. Сделаны выводы.

In [182]:
# for i in range(1, 13):
#     at = splitting_vectors(i)
#     bt = splitting_vectors(2**i - 1)
#     print(i)
#     chaina = star_chain(at)
#     chainb = star_chain(bt)
#     print("Гипотеза верна: ", len(chainb) <= i - 1 + len(chaina), len(chainb),len(chaina))

1

Гипотеза верна:  True 1 1

2

Гипотеза верна:  True 3 2

3

Гипотеза верна:  True 5 3

4

Гипотеза верна:  True 6 3

5

Гипотеза верна:  True 8 4

6

Гипотеза верна:  True 9 4

7

Гипотеза верна:  True 11 5

8

Гипотеза верна:  True 11 4

9

Гипотеза верна:  True 13 5

10

Гипотеза верна:  True 14 5

11

Гипотеза верна:  True 16 6

12

Гипотеза верна:  True 16 5

| n  |  l(n) | l(2^n-1)  |  Проверка |
|---|---|---|---|
| 1  | 1  | 1  | Гипотеза верна  |
| 2  | 2  | 3  | Гипотеза верна  |
| 3  | 3  | 5  | Гипотеза верна  |
| 4  | 3  | 6  | Гипотеза верна  |
| 5  | 4  | 8  | Гипотеза верна  |
| 6  | 4  | 9  | Гипотеза верна  |
| 7  | 5  | 11  | Гипотеза верна  |
| 8  | 4  | 11  | Гипотеза верна  |
| 9  | 5  | 13  | Гипотеза верна  |
| 10  | 5  | 14  | Гипотеза верна  |
| 11  | 6  | 16  | Гипотеза верна  |
| 12  | 5  | 16  | Гипотеза верна  |

## Выводы

Общий вывод по проделанной работе.